ASoC: SOF: ipc4: Support for mtrace, log extraction from SRAM debug window for TGL/MTL

ujfalusi opened the pull request:
Hi,

This series will enable the log extraction via debug window on an IPC4 firmware which have support for this mode of logging.

The initial implementation supports the type of logging used by TGL and MTL.

The logs are generated per core into a separate slot and the initial implementation follows that, but if desired we can make a single file to mix the logs from all cores (which might result with jumping timestamps).
The parameters for the log configuration in firmware also exposed for the user to adjust.

Péter

kv2019i commented on sound/soc/sof/ipc4.c:
```diff
@@ -545,6 +545,8 @@ static int ipc4_fw_ready(struct snd_sof_dev *sdev, struct sof_ipc4_msg *ipc4_msg
 	sdev->host_box.offset = outbox_offset;
 	sdev->host_box.size = outbox_size;
 
+	sdev->debug_box.offset = snd_sof_dsp_get_window_offset(sdev, 2);
```
Add a define/constant for the magic value?

ujfalusi replied:
Hrm, the only define came to mind would be `SOF_DSP_WINDOW_2`
In other places we also ask for the window index by nimber, I think. If not, then I'll yry to be more creative ;)

ujfalusi replied:
I'll use `SOF_IPC4_DEBUG_WINDOW_IDX` and add `SOF_IPC4_OUTBOX_WINDOW_IDX` for the other window we are asking for with index.

kv2019i commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,474 @@
+	if (core_data->host_read_ptr != core_data->dsp_write_ptr)
+		return TRUE;
+
+	return false;
```
TRUE on L113 and false on L115 looks weird. I'd more used to lower-case, but at least it should be the same.

ujfalusi replied:
Oops, this slipped through, thanks for spotting it!

kv2019i reviewed (changes requested):
Almost perfect! :) A few minor nits, but looks good to go otherwise.

ujfalusi commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,474 @@
+			"%s: core%d, host read: %#x, dsp write: %#x, avail: %#x",
+			__func__, core_data->id, read_ptr, write_ptr, avail);
+
+	log_data = kmalloc(avail, GFP_KERNEL);
```
the lack of `kfree` is striking :o
/me hides

RanderWang replied:
dynamical alloc & free ? why not create a 4k size buffer in trace initialization function and free it when trace is free ?

ujfalusi replied:
or do the allocation in the core file open and free it up when closed.

yongzhi1 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,474 @@
+	}
+
+	if (!ipc4_data->mtrace_log_bytes ||
+	    ipc4_data->mtrace_type != SOF_IPC4_MTRACE_INTEL_TGL) {
```
Is mtrace_log_bytes a fixed number? what is the use of it's value? thanks,

ujfalusi replied:
It is supposed to be the size of the debug window, but on TGL it is coming back as the size of _one_ slot in the window, on APL it is 0x2000 (2 cores, 2 slots, I think).
If it is 0 then we conclude that the firmware does not supports mtrace type of logging (this is the case atm with SOF.IPC4).
On TGL type of platforms we _know_ that the trace window have 16 slots, it is static, so if the number is not NULL, mtrace can be enabled.

RanderWang commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -190,6 +191,7 @@ static int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 			break;
 		case SOF_IPC4_FW_CFG_TRACE_LOG_BYTES:
 			trace_sof_ipc4_fw_config(sdev, "Trace log size", *tuple->value);
+			ipc4_data->mtrace_log_bytes = *tuple->value;
```
@yongzhi1  this value is not correct in your test ?

yongzhi1 replied:
In my test, the code does not hit line 192, seems there is no SOF_IPC4_FW_CFG_TRACE_LOG_BYTES defined in FW configuration.

ujfalusi replied:
@yongzhi1, is it with SOF firmware? It is not yet supported there and this is the reason why it is checked against.

RanderWang commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -13,18 +13,27 @@
 
+enum sof_ipc4_mtrace_type {
+	SOF_IPC4_MTRACE_NOT_AVAILABLE = 0,
+	SOF_IPC4_MTRACE_INTEL_TGL,
```
the trace design is based on CAVS version. This PR is applicable to cavs2x+ platform, such as icelake and tigerlake, mtl. For SKL, APL, CML, we need cavs1x.  Can we rename it to SOF_IPC4_MTRACE_INTEL_CAVS2X ?

ujfalusi replied:
@RanderWang, OK, let's change to `SOF_IPC4_MTRACE_INTEL_CAVS2X`.
As a sidenote: it is unclear how this maps for me: it looks like that APL and CNL have incompatible logging and it depends where you are looking, SKL is either uses APL or CNL type of mtrace.
CAVS1x will not cut it as we have at least two types on there.
But let's care about that if we ever implement support for those...

RanderWang replied:
yes,  we should use CAVS15 APL and CAVS18 CNL,

ujfalusi replied:
Thanks, I think the only difference is the message itself to enable the logging. library vs core and APL does not have *_timer_period parameters while CNL have.
I'll still left them out until it is clear and tested.

ujfalusi commented:
Changes since v1:
- Use define for the debug window index (and add define for the outbox window)
- Allocate the temp buffer for the log data on file open and free it up on close
- rename SOF_IPC4_MTRACE_INTEL_TGL to SOF_IPC4_MTRACE_INTEL_CAVS2X
- Use kernel style boolean name (s/TRUE/true)

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+#include "ipc4-priv.h"
+
+/*
+ * debug info window is orginized in 16 slots.
```
typo organized

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+{
+	struct sof_mtrace_core_data *core_data = file->private_data;
+	struct snd_sof_dev *sdev = core_data->sdev;
+	u32 log_buffer_offset = core_data->slot_offset + 8;
```
that comment /* byt 0 - core ID */ is more useful here than in the header file

ujfalusi replied:
It is not applicable here. This is the log buffer where the first u32 is the host_read_ptr, the second is the dsp_write_ptr...

ujfalusi replied:
I have moved the assignments down and added comments to explain the `+ 8` (which I changed to `sizeof(u32) * 2`)

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+	int i;
+
+	if (!sdev)
+		return -EINVAL;
```
is this check really required? With the function being static and where it is called, sdev cannot possibly be NULL right?

ujfalusi replied:
True, dropped.

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+			   &priv->state_info.aging_timer_period);
+	debugfs_create_u32("fifo_full_timer_period", 0644, dfs_root,
+			   &priv->state_info.fifo_full_timer_period);
+	/* Separate priorities mask file per library, index 0 is basefw */
```
what is the intented use for separate priorities? And how are we verifying this?

ujfalusi replied:
The use case is that user can configure the priority for each library in case it is needed, like if we are not interested info prints from basefw, but we want to see them from the library/module under debugging.
We don't verify it, let me see if this can even be library/module dependent, in that case we can not verify this.

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+	msg.data_ptr = &priv->state_info;
+	if (iops->set_get_data(sdev, &msg, msg.data_size, true)) {
+		sdev->fw_trace_is_supported = false;
+		return 0;
```
this is confusing to return 0 in both cases isnt it? How ill we ever know if trace is enabled or not?

ujfalusi replied:
There will be an error print from IPC level and we will disable the mtrace as it is not working, but we don't want to fail the whole audio driver load or see these errors again.

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+	msg.data_size = sizeof(priv->state_info);
+	msg.data_ptr = &priv->state_info;
+	if (iops->set_get_data(sdev, &msg, msg.data_size, true))
+		sdev->fw_trace_is_supported = false;
```
not sure I follow the logic @ujfalusi if the IPC to disable logging fails, why should we set this flag to false?

ujfalusi replied:
OK, I'll drop the flag setting, we should not be here if the mtrace is not supported by the firmware, we would have failed earlier in enable.

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MOD_ID(SOF_IPC4_MOD_INIT_BASEFW_MOD_ID);
+	msg.primary |= SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID);
+	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_ENABLE_LOGS);
```
how do we distinguish enable from disable in the IPC? the IPC looks the same to me for both or am I going blind?

ujfalusi replied:
The IPC message is the same and to enable/disable the logging is down within the payload of the message, at line 348 and 318

ranj063 commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,509 @@
+	int i;
+
+	slot_desc_offset = sdev->debug_box.offset;
+	for (i = 0; i < MAX_MTRACE_SLOTS - 1; i++) {
```
-1 because?

ujfalusi replied:
Because we have 16 slots (MAX_MTRACE_SLOTS), the first slot is used for descriptors to describe the remaining 15 slots (MAX_MTRACE_SLOTS - 1).
I have added a comment.

ranj063 commented on sound/soc/sof/intel/mtl.c:
```diff
@@ -771,6 +771,8 @@ int sof_mtl_ops_init(struct snd_sof_dev *sdev)
 	ipc4_data = sdev->private;
 	ipc4_data->manifest_fw_hdr_offset = SOF_MAN4_FW_HDR_OFFSET;
 
+	ipc4_data->mtrace_type = SOF_IPC4_MTRACE_INTEL_CAVS2X;
```
Its awkward to call it MTRACE_INTEL_CAVS2X for MTL. Can we find a better name that applies to both CAVS2X and ACE?

plbossart replied:
indeed this really confusing naming. MTL is NOT CAVS2x

ujfalusi replied:
But MTL (which is ACE) is using the same method as it was introduced by CAVS2X, so mtrace in MTL is MTRACE_INTEL_CAVS2X.
The CAVS2X mtrace is using per module/library log masking (up to 16 lib is supported), previous mtrace for CAVS1.5, 1.8 had par core enable and priority, with 1.8 having the timer_period parameters..

Imho the mtrace type in MTL is CAVS2X.

kv2019i replied:
Agree with @ujfalusi . Other option is to use arbitrary versions, but those would be a version defined only in Linux driver, so I'd say potential for even more confusion. The "platform where this was introduced" naming concept seems least bad.

plbossart replied:
why can't we list all known types of MTRACE today, and only add support for what we tested. Adding definitions like this seems very ad-hoc and not based on architecture/firmware evolutions.

ujfalusi replied:
If we don't need to document the differences, then OK and there is this fourth version I have seen bugs me a lot.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -427,6 +427,11 @@ struct sof_ipc4_dx_state_info {
+#define SOF_IPC4_LOG_CORE_MASK			GENMASK(15, 12)
+#define SOF_IPC4_LOG_CORE_GET(x)		(((x) & SOF_IPC4_LOG_CORE_MASK) >> \
+						 SOF_IPC4_LOG_CORE_SHIFT)
+
```
the commit message is rather confusing.

Is this saying we have one log buffer per core? This isn't something we had before, so needs more explanations on how the logs are used.

ujfalusi replied:
@plbossart, yes, we have one log buffer per core.
I'll try to update the commit message to convey this better.

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -13,6 +13,9 @@
 #include <sound/sof/ext_manifest4.h>
 #include "sof-priv.h"
 
+/* The DSP window indexes are fixed */
```
typo: index or indices, pick one.

ujfalusi replied:
OK, I'll use indices

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -545,6 +545,9 @@ static int ipc4_fw_ready(struct snd_sof_dev *sdev, struct sof_ipc4_msg *ipc4_msg
 
+	sdev->debug_box.offset = snd_sof_dsp_get_window_offset(sdev,
+							SOF_IPC4_DEBUG_WINDOW_IDX);
+
```
should there be an added log below for symmetry with the dsp_box and host_box?

ujfalusi replied:
I will log the debug window start, thanks for the suggestion.

plbossart commented on sound/soc/sof/Makefile:
```diff
@@ -9,7 +9,8 @@ snd-sof-objs +=	ipc3.o ipc3-loader.o ipc3-topology.o ipc3-control.o ipc3-pcm.o\
 ifneq ($(CONFIG_SND_SOC_SOF_INTEL_IPC4),)
-snd-sof-objs += ipc4.o ipc4-loader.o ipc4-topology.o ipc4-control.o ipc4-pcm.o
+snd-sof-objs += ipc4.o ipc4-loader.o ipc4-topology.o ipc4-control.o ipc4-pcm.o\
+		ipc4-mtrace.o
```
commit message

One of the debugging/logging features

Also you need to explain how the logs are used/collected when there are multiple cores, same feedback as in previous patch.

ujfalusi replied:
OK

kv2019i reviewed (approved):
All my concerns addressed!

ujfalusi commented:
Changes since v2:
- Comments updated and added in place to explain the code
- typos fixed
- commit messages updated with more direct wording
- logs_priorities_mask representation changed to a single file
  - reading the file will print all 16 masks
  - writing a tuple ( id,mask) would update the given mask
- fixes for position update race and read pointer update
- update the use buffer with 0 bytes returned when there is no new data available (happens when closing the file while the DSP is suspended)

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+ * The two pointers are offsets within the buffer.
+ */
+
+#define FW_EPOCH_DELTA			11644473600LL
```
What does this value represent? Is this tied to a specific frequency or operating mode.
A comment would not hurt.

ujfalusi replied:
I have the comment at line 396 where this is used: `	/* The system time is in usec, UTC, epoch is 1601-01-01 00:00:00 */`
Magic number to convert from POSIX UTC time to NT time, needed by the firmware.

plbossart replied:
I don't know what NT time is

ujfalusi replied:
The firmware specification tells that the EPOCH for it's time is 1601-01-01 00:00:00, UTC. 
ktime_get_real() will return the time since UNIX/POSIX EPOCH (1970-01-01 00:00:00, UTC).

I don't know what happens if we just pass the Unix time, probably nothing serious that user space tooling can not handle?

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+			core_data->slot_offset = sdev->debug_box.offset;
+			core_data->slot_offset += SOF_MTRACE_SLOT_SIZE * (i + 1);
+			dev_dbg(sdev->dev, "%s: slot%d is used for core%u\n",
+				__func__, i, core);
```
should this be i+1, if slot 0 is used for descriptors it cannot be used for a specific core, can it?

It would be simpler to start the loop at 1 really!
for (i = 1; i < MAX_MTRACE_SLOTS; i++) {

ujfalusi replied:
> should this be i+1, if slot 0 is used for descriptors it cannot be used for a specific core, can it?

The first usable slot as a real slot is the second, that is true. The debug window is split up to equal sized chunks, mostly referred as slots. The first one is the descriptor are. In a way it is a slot but from another angle it is not really a slot as it can not be used for anything else than for the descriptors.
The first (index 0) descriptor at the start of the debug window is describing the slot at debug_window + slot_size, since the descriptor are is made to be the same size as the other slots.

> It would be simpler to start the loop at 1 really! for (i = 1; i < MAX_MTRACE_SLOTS; i++) {

Depending on how we are looking at this, yes, but for me, when I look at for (i = 1; ...) my brain will say first that oh, we are skipping the first entry in this loop. But we don't. We read the 0th descriptor as well, which describes the first slot _after_ the descriptor area.

plbossart replied:
right, but the message is not correct. Core0 uses slot1.

ujfalusi replied:
So, I have re-checked the sources and in the firmware we have `#define MEMORY_WINDOW_SLOTS_COUNT   15`, so the descriptor area is not counted as a slot.
The `MAX_MTRACE_SLOTS = 16` was inherited from an earlier implementation from @bardliao / @RanderWang and I kept it as it was.

Now changed the code to match with how the firmware refers to slots and that will 'solve' the +1 -1 confusion.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+				__func__, i, type);
+		}
+
+		slot_desc_offset += sizeof(u32) * 3;
```
3 is magic number, better use a define.

ujfalusi replied:
At line 16 I have the information on the descriptor layout, which is three u32 per slot, `*_SLOT_DESCRIPTOR_SIZE 12` would be better?

plbossart replied:
sounds good.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+			}
+		} else if (type) {
+			dev_dbg(sdev->dev, "%s: slot%d is not a log slot (%#x)\n",
+				__func__, i, type);
```
same, is this i+1?

ujfalusi replied:
Again, depends on how we look at things. Do we count the descriptor area as real slot?

plbossart replied:
dunno. The point in that you will not check slot15 with your loop....

ujfalusi replied:
This is the organization of the memory at the debug offset:
```
------------------------
| Page0  - descriptors |
------------------------
| Page1  - slot0       |
------------------------
| Page2  - slot1       |
------------------------
|         ...          |
------------------------
| Page14  - slot13     |
------------------------
| Page15  - slot14     |
------------------------
```
and page size == slot size
I have now changed the code and comments to describe this.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+
+	/* default enable default trace state_info */
+	priv->state_info.aging_timer_period = 10;
+	priv->state_info.fifo_full_timer_period = 10;
```
magic numbers, and units not specified.

ujfalusi replied:
OK, I'll add something `*_DEFAULT_TIMER_PERIOD 10` for these and try to find the unit, I can not recall seeing it, but it should be somewhere.

ujfalusi replied:
They are in ms

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+
+	ret = mtrace_debugfs_create(sdev);
+	if (ret < 0)
+		return ret;
```
disable the trace on errors?

ujfalusi replied:
Yes! Thanks for spotting it, last minute change to not create the debugfs if enable fails.

ujfalusi replied:
Fixed

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -17,18 +17,27 @@
 
+enum sof_ipc4_mtrace_type {
+	SOF_IPC4_MTRACE_NOT_AVAILABLE = 0,
+	SOF_IPC4_MTRACE_INTEL_CAVS2X,
```
I still don't get this definition, can we not list all the possible traces today, and later add support for 1.5 and 1.8.
We should describe everything we know, not add a random value just because we tested this on TGL only.

ujfalusi replied:
I'm aware of those two other variations, but I have seen a third one in some place, but can not connect it to anything.
But if we add them, should we document them as well? They are different than similar to the one we implement here.

plbossart replied:
no, just like we added variants of cAVS architectures, we can add variants of the trace. We don't need to document anything, just be consistent on which machine uses which version of the trace.

ujfalusi replied:
@plbossart, I think the first PR had those, but I have named them APL, CNL and TGL type, let me find that and update the definitions and add some more checks in the code.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -589,6 +589,10 @@ static void sof_ipc4_rx_msg(struct snd_sof_dev *sdev)
 		break;
+	case SOF_IPC4_NOTIFY_LOG_BUFFER_STATUS:
+		sof_ipc4_mtrace_update_pos(sdev,
+					   SOF_IPC4_LOG_CORE_GET(ipc4_msg->primary));
```
one line?

ujfalusi replied:
OK, I can live with 92 column

ujfalusi commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,607 @@
+
+	sdev->fw_trace_data = priv;
+
+	/* default enable default trace state_info */
```
haha, nice comment ;)

ujfalusi commented:
Changes since v3:
- Description of the memory layout in the debug window is updated: the first page is no longer referred as slot as it is a descriptor area, the number of slots are 15.
- Added definitions to other type of mtrace implementations (cavs 1.5 and 1.8)
- Error handling in init reworked to be cleaner and `ipc4_mtrace_enable()` now returns error codes
- Comments updated for clarity
- defines added for magic numbers.

RanderWang commented on sound/soc/sof/intel/tgl.c:
```diff
@@ -80,6 +80,8 @@ int sof_tgl_ops_init(struct snd_sof_dev *sdev)
 		ipc4_data = sdev->private;
 		ipc4_data->manifest_fw_hdr_offset = SOF_MAN4_FW_HDR_OFFSET;
 
+		ipc4_data->mtrace_type = SOF_IPC4_MTRACE_INTEL_CAVS_2;
```
icl.c ?   also CAVS_2

ujfalusi replied:
@RanderWang, the icl.c did not had the `if (sdev->pdata->ipc_type == SOF_INTEL_IPC4) { ... }` section, so it is not made to work with IPC4. I think it can use the same section as tgl?

Should I add that as a separate patch first?

ujfalusi commented:
Changes since v4:
- Add missed IPC4 support for ICL
  - set mtrace version for ICL as CAVS_2 (thanks @RanderWang for pointing out)
- added missing log source definitions.

RanderWang reviewed (dismissed):
Well done!

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -427,6 +427,11 @@ struct sof_ipc4_dx_state_info {
+#define SOF_IPC4_LOG_CORE_MASK			GENMASK(15, 12)
+#define SOF_IPC4_LOG_CORE_GET(x)		(((x) & SOF_IPC4_LOG_CORE_MASK) >> \
+						 SOF_IPC4_LOG_CORE_SHIFT)
+
```
nit-pick on commit message

ID of the core which updated its log buffer.

ujfalusi replied:
OK

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,647 @@
+	if (sof_debug_check_flag(SOF_DBG_PRINT_DMA_POSITION_UPDATE_LOGS))
+		dev_dbg(sdev->dev,
+			"%s: core%d, host read: %#x, dsp write: %#x, avail: %#x\n",
+			__func__, core_data->id, read_ptr, write_ptr, avail);
```
we want to avoid using __func__ in dev_dbg

ujfalusi replied:
Right, this was done _before_ we started to drop these.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,647 @@
+		return -ENOMEM;
+
+	for (i = 0; i < MAX_ALLOWED_LIBRARIES; i++) {
+		offset = strlen(buf);
```
I don't get this code. buf is not modified in this loop, so offset is constant.

ujfalusi replied:
We are printing into the buffer at offset position and offset is the position of the string terminator (`0`).
i = 0 -> offset = 0
i =1 -> offset = 15 (first line: 14 char + new line)
...
We just place the lines after each other, we could use a define for 15 and use that one as a base for the offset, but I rather keep this as it is.

plbossart replied:
my point was that this line could have been moved out of the loop @ujfalusi

ujfalusi replied:
No, after each iteration the strlen(buf); will increase, the `strlen` of the buf changes since we put lines in there.

plbossart replied:
oh, I see. I completely missed that part.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,647 @@
+
+			if (core >= sdev->num_cores) {
+				dev_dbg(sdev->dev, "%s: core%u is invalid for slot%d\n",
+					__func__, core, i);
```
remove __func__

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,647 @@
+			core_data->slot_offset = sdev->debug_box.offset;
+			core_data->slot_offset += SOF_MTRACE_SLOT_SIZE * (i + 1);
+			dev_dbg(sdev->dev, "%s: slot%d is used for core%u\n",
+				__func__, i, core);
```
__func__ again

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,647 @@
+
+	if (sof_debug_check_flag(SOF_DBG_PRINT_DMA_POSITION_UPDATE_LOGS))
+		dev_dbg(sdev->dev, "%s: core%d, host read: %#x, dsp write: %#x",
+			__func__, core, core_data->host_read_ptr,
```
__func__

plbossart reviewed (changes requested):
couple of nit-picks and one doubt on the buff/offset computation.

ujfalusi commented:
Changes since v5:
- drop `__func__` from dev_dbg() prints.

RanderWang commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -13,18 +13,33 @@
 #include "sof-priv.h"
 
+/* The DSP window indices are fixed */
+#define SOF_IPC4_OUTBOX_WINDOW_IDX	1
```
we can also add "SOF_IPC4_INBOX_WINDOW_IDX 0"

ujfalusi replied:
We always get the inbox offset with the call `snd_sof_dsp_get_mailbox_offset()`, so I thought it would be just awkward to have a new define, which is unused?

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,642 @@
+struct sof_mtrace_core_data {
+	struct snd_sof_dev *sdev;
+
+	int id;
```
is this the core id?

ujfalusi replied:
Yes, I opted to not prefix it with `core_` since it is part of the `core_data` struct.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,642 @@
+	void *log_buffer;
+	u32 host_read_ptr;
+	u32 dsp_write_ptr;
+	bool missed_update;
```
a comment on what this is wouldn't hurt - no idea how this might work.

ujfalusi replied:
OK, I'll rename it as well to pending_pos_update.
It is set in case a position update notification arrives before we would be able to get the information about the slots.
We can not read the position from the slot if the core does not have the slot information up to date.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,642 @@
+	u32 slot_offset;
+	void *log_buffer;
+	u32 host_read_ptr;
+	u32 dsp_write_ptr;
```
is there a size or something allowing to detect wrap-around for long usages?

ujfalusi replied:
These pointers are wrapping within the size of the slot, they will be never larger than `0x1000 - 2 * sizeof(u32)`

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,642 @@
+	enum sof_mtrace_state mtrace_state;
+	struct sof_log_state_info state_info;
+
+	struct sof_mtrace_core_data cores[];
```
shouldn't there be a core_count or something? Or is this taken from the chip_info descriptor?

ujfalusi replied:
It is dynamically allocated based on the number of cores the booted system have.

plbossart commented on sound/soc/sof/ipc4-mtrace.c:
```diff
@@ -0,0 +1,642 @@
+	if (unlikely(ret))
+		return ret;
+
+	core_data->log_buffer = kmalloc(SOF_MTRACE_SLOT_SIZE, GFP_KERNEL);
```
should the log_buffer be zeroed out?

ujfalusi replied:
not really. We are always writing to it first then copy the same amount out to user space.

plbossart commented on sound/soc/sof/intel/mtl.c:
```diff
@@ -770,6 +770,8 @@ int sof_mtl_ops_init(struct snd_sof_dev *sdev)
 	ipc4_data = sdev->private;
 	ipc4_data->manifest_fw_hdr_offset = SOF_MAN4_FW_HDR_OFFSET;
 
+	ipc4_data->mtrace_type = SOF_IPC4_MTRACE_INTEL_CAVS_2;
```
we're not testing if mtrace_type is supported anytwhere?

ujfalusi replied:
yes, we do in `ipc4_mtrace_init()`:
```c
	if (!ipc4_data->mtrace_log_bytes ||
	    ipc4_data->mtrace_type != SOF_IPC4_MTRACE_INTEL_CAVS_2) {
		sdev->fw_trace_is_supported = false;
		return 0;
	}
```
Only CAV_2 is supported.

ujfalusi commented:
Changes since v6:
- missed_update flag renamed to delayed_pos_update

plbossart reviewed (approved):
No further questions your Honor.

RanderWang reviewed (approved):
LGTM

plbossart commented:
@ranj063 we could use your review on this one too.

ranj063 reviewed (approved):
looks great @ujfalusi
